### PR TITLE
[AIRFLOW-4106][AIRFLOW-4422] Add pool instrumentation

### PR DESCRIPTION
**Pull request by milton0825**
Cherry-pick https://github.com/apache/airflow/pull/5453 and https://github.com/apache/airflow/pull/4927 from upstream. Our branch actually diverged a lot from the upstream so I end up just writing the code myself.

Added a few stats around Airflow pool
```
pool.open_slots.<pool_name> | Number of open slots in the pool
pool.used_slots.<pool_name> | Number of used slots in the pool
pool.starving_tasks.<pool_name> | Number of starving tasks in the pool
```

### CODE REVIEW

**Comment by milton0825** [comment]
@astahlman @feng-tao @ArgentFalcon @youngyjd

File: airflow/jobs.py [hunk 1]
```diff
@@ -1024,35 +1024,39 @@ def _execute_task_instances(self,
...
             dag_id_to_possibly_running_task_count = {}
 
-            for task_instance in priority_sorted_task_instances:
+            # Number of tasks that cannot be scheduled because of no open slot in pool
+            num_starving_tasks = 0
+            for current_index, task_instance in enumerate(priority_sorted_task_instances):
                 if open_slots <= 0:
                     self.logger.info("No more slots free")
                     # Can't schedule any more since there are no more open slots.
+                    num_starving_tasks = len(priority_sorted_task_instances) - current_index
```
Comment (astahlman) [inline airflow/jobs.py]:
```suggestion
                    num_starving_tasks += 1
```

Would this be simpler?

Comment (milton0825) [reply airflow/jobs.py]: The reason we need this is because we break the loop immediately when there are no more `open_slots`.

Comment (astahlman) [reply airflow/jobs.py]: Ohhh, I see. Thanks

**Comment by feng-tao** [comment]
how about the default pool?

File: airflow/jobs.py [hunk 2]
```diff
@@ -1024,35 +1024,39 @@ def _execute_task_instances(self,
                 # Arbitrary:
                 # If queued outside of a pool, trigger no more than
                 # non_pooled_task_slot_count per run
-                open_slots = conf.getint('core', 'non_pooled_task_slot_count')
+                slots = conf.getint('core', 'non_pooled_task_slot_count')
+                open_slots = slots
+                pool_name = 'default_pool'
```
Comment (milton0825) [inline airflow/jobs.py]: @feng-tao default pool is captured here

**Pull request merged**